Fix Laravel 5.4 support

tortuetorche opened the pull request:
Fix the `BindingResolutionException`:
```
 in Container.php line 804: Target [Symfony\Component\HttpKernel\HttpKernelInterface] is not instantiable
```

Because, since Laravel 5.4:
> the container's make method no longer accepts a second array of parameters.

We also can use this line of code, but it's only compatible with PHP >= 5.6.0:
```php
$middleware = new $callable(...$params);
```
Source: http://stackoverflow.com/a/2409288

tortuetorche committed:
Fix Laravel 5.4 support

Since Laravel 5.4:
> the container's make method no longer accepts a second array of parameters.

barryvdh commented:
We could wait for this to be tagged: https://github.com/laravel/framework/commit/5d9b363448b9aac83b610166141b8bba9f1807f8
which brings back `makeWith` or `app($instance, $parameters)`

tortuetorche commented:
Oh, I wasn't aware of this `makeWith()` method.
Thank you Barry and have a good afternoon 👍

tortuetorche commented:
Laravel [5.4.16](https://github.com/laravel/framework/tree/v5.4.16) is released!
So to keep Laravel < 5.4 compatibility, we can do something like:
```php
$makeMethod = method_exists($this->container, 'makeWith') ?  'makeWith' : 'make';
$middleware = $this->container->$makeMethod($callable, $params);
```
@barryvdh What do you think about this?

barryvdh commented:
Sounds good

tortuetorche committed:
Use the Container::makeWith() method if present

Since Laravel 5.4.16, the Container::makeWith($class, $params) method replace the old behavior of Container::make() method.

barryvdh commented:
Ready to go?

tortuetorche commented:
No, I need to do some manual tests...

tortuetorche commented:
The `Container::makeWith()` method doesn't work as expected and I don't know why yet.

tortuetorche commented:
I successfully call the `Container::makeWith()` method with some modifications:

In `vendor/barryvdh/laravel-stack-middleware/src/StackMiddleware.php` file:
```php          
            // Add kernel as 'app' parameter
            $params['app'] = $kernel;
            $makeMethod = method_exists($this->container, 'makeWith') ?  'makeWith' : 'make';
            $middleware = $this->container->$makeMethod($callable, $params);
```

And in my Turbolinks Service Provider [`vendor/efficiently/turbolinks/src/Frenzy/Turbolinks/TurbolinksServiceProvider.php`](https://github.com/efficiently/turbolinks/blob/dev/src/Frenzy/Turbolinks/TurbolinksServiceProvider.php), I replace this code:
```php
        $stack->bind(
            'Frenzy\Turbolinks\Middleware\StackTurbolinks',
            'Helthe\Component\Turbolinks\StackTurbolinks',
            [$this->app['turbolinks']]
        );
```
By this one:
```php
        $stack->bind(
            'Frenzy\Turbolinks\Middleware\StackTurbolinks',
            'Helthe\Component\Turbolinks\StackTurbolinks',
            ['turbolinks' => $this->app['turbolinks']]
        );
```
See [`StackTurbolinks::__construct()`](https://github.com/helthe/Turbolinks/blob/2.0.0/StackTurbolinks.php#L40) method, for more information.

The `$params` parameter, of `Container::makeWith()` method, need to have named keys, but I don't know why...

tortuetorche commented:
I've open an issue on the Laravel Framework repository, see: https://github.com/laravel/framework/issues/18474

tortuetorche commented:
According to @themsaid https://github.com/laravel/framework/issues/18474#issuecomment-289062443,
> With `Container::makeWith()` method, you have to pass the parameter name.

So it's a breaking change compared to the behavior of `Container::make()` of Laravel 5.3.
Because instead of binding Laravel Stack Middleware, like this:
```php
        $stack->bind(
            'Frenzy\Turbolinks\Middleware\StackTurbolinks',
            'Helthe\Component\Turbolinks\StackTurbolinks',
            [$this->app['turbolinks']]
        );
```
We need to bind Laravel Stack Middleware, like that:
```php
        $stack->bind(
            'Frenzy\Turbolinks\Middleware\StackTurbolinks',
            'Helthe\Component\Turbolinks\StackTurbolinks',
            ['turbolinks' => $this->app['turbolinks']]
        );
```

barryvdh commented:
OK can you update the pr?

tortuetorche committed:
With Container::makeWith() method, we have to pass the parameter name.

https://github.com/laravel/framework/issues/18474

tortuetorche commented:
@barryvdh Done, but these breaking changes should be documented in the README.md

barryvdh commented:
Is it breaking?

tortuetorche commented:
When I said "it's a breaking change", I mean, you can't do this anymore:
```php
app('stack')->bind('AddRobotsHeaders', 'League\StackRobots\Robots', ['production', 'APP_ENV']);
```
Instead, you need to do this:
```php
app('stack')->bind('AddRobotsHeaders', 'League\StackRobots\Robots', ['env' => 'production', 'envVar' => 'APP_ENV']);
```
